### PR TITLE
Split cmake starter workflow into single/multi platform support

### PR DESCRIPTION
**Pull request by jgoshi**
Rename the existing starter workflow to indicate that it supports a single platform. Add a new starter workflow that has cross-platform support.

## Pre-requisites

- [x] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least privileged [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully.

**For _CI_ workflows, the workflow:**

- [x] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [x] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [NA] Packaging workflows should run on `release` with `types: [ created ]`.
- [NA] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [NA] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/code-scanning).
- [NA] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [NA] `name`: Name of the Code Scanning integration.
  - [NA] `creator`: Name of the organization/user producing the Code Scanning integration.
  - [NA] `description`: Short description of the Code Scanning integration.
  - [NA] `categories`: Array of languages supported by the Code Scanning integration.
  - [NA] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [NA] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [x] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [x] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [x] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

File: ci/cmake-single-platform.yml
Comment (michaelbprice) [inline ci/cmake-single-platform.yml:1]: I'm wondering if we shouldn't be more explicit and name this `cmake-ubuntu.yml`. I guess a possible determining factor would be the answer to:

How easily could take this single platform script and convert it to run on a different platform?

File: ci/cmake-single-platform.yml
Comment (michaelbprice) [inline ci/cmake-single-platform.yml:3]: Along the same lines as the naming issue, maybe just `CMake on Ubuntu`?  There's a tradeoff:

* `CMake on Ubuntu` - The user gets a clear picture from the GitHub.com UI that the workflow is for Ubuntu
VS
* `CMake on a single platform` - The user won't need to make changes to the name if they change the nature of the workflow (or copy it to make a separate Windows or Mac workflow file). This also avoids the risk of having an "incorrect" name.

File: ci/properties/cmake-multi-platform.json
Comment (michaelbprice) [inline ci/properties/cmake-multi-platform.json:2]:
```suggestion
    "name": "CMake based, multi-platform projects",
```

File: ci/properties/cmake-single-platform.json
Comment (michaelbprice) [inline ci/properties/cmake-single-platform.json:2]:
```suggestion
    "name": "CMake based, single-platform projects",
```

File: ci/properties/cmake-single-platform.json
Comment (michaelbprice) [inline ci/properties/cmake-single-platform.json:1]: This file and its contents are also subject to the naming question.

File: ci/properties/cmake-multi-platform.json
Comment (michaelbprice) [inline ci/properties/cmake-multi-platform.json:4]: @andrewakim Do we know what this `iconName` field controls? If it's used to select an icon, should we consider having some sort of adorned icon to distinguish them?

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:41]: Is this style of putting the comment after the first field of the step consistent with other workflow files? It feels strange to me personally.

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:16]: Should we use column-limit for the contents of the file, particularly for comments and for longer commands?

@andrewakim - Do you know if there's a GitHub Actions YAML formatting action that could just apply some standards across starter workflow file contents (and other workflows if others wanted to use it).

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:45]: One of the things that I've noticed within GitHub Actions documentation is that workflow expressions are usually formatted a particular way, namely instead of
```
${{github.workspace}}
```
You see (with leading and trailing whitespace)
```
${{ github.workspace }}
```

@andrewakim - Is this a style thing that we should reproduce here as well?

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:50]: Even if we don't enforce a column width limit, I think we should make this a multi-line run step. It's the one that users are most likely to modify/extend (an assumption I'm making), so it's the one they'll most likely be tempted to make multi-line, which if you don't already know how, can be frustrating.

```suggestion
      run: |
        cmake -B ${{ steps.strings.outputs.build-output-dir }} \
        -DCMAKE_CXX_COMPILER=${{matrix.compiler}} \
        -DCMAKE_C_COMPILER=${{matrix.compiler}} \
        -DCMAKE_BUILD_TYPE=${{matrix.build_type}} \
        -S ${{github.workspace}}
```

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [reply ci/cmake-multi-platform.yml:60]: We could probably do similar things to the contents of the single platform workflow, but let's avoid that for now so that we can avoid having multiple variables while we try to understand if users are finding the multi-platform workflow a better experience.

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:50]: `-DCMAKE_CXX_COMPILER=${{matrix.compiler}} -DCMAKE_C_COMPILER=${{matrix.compiler}}` is not quite correct. For `cl` it is, but for GCC and Clang, the correct value to use for `CMAKE_CXX_COMPILER` are the `g++` and `clang++` frontend driver programs, repsectively.

**Review by michaelbprice — CHANGES_REQUESTED** [review]
I would like to see at least 1, perhaps 2, examples of this workflow applied to and working well for some example C++ code projects. If you need suggestions on ones to try, I can hunt some down. That would also let us fork and experiment with what the experience is to take these workflows and extend them for additional cases (like adding Mac support, adding in a dependency management step, or moving to Ninja as a generator).

File: ci/cmake-single-platform.yml
Comment (jgoshi) [reply ci/cmake-single-platform.yml:1]: I think it's straightforward to just change the os. I think maybe we can leave the name for now?

File: ci/cmake-multi-platform.yml
Comment (jgoshi) [reply ci/cmake-multi-platform.yml:50]: You mean in the matrix I should use g++ and clang++ instead of gcc and clang?

Comment (michaelbprice) [reply ci/cmake-multi-platform.yml:50]: More complex than that. An example invocation of cmake should look like this:
```
cmake -B ./build -DCMAKE_CXX_COMPILER=g++ -DCMAKE_C_COMPILER=gcc -DCMAKE_BUILD_TYPE=Release -S .
```

Comment (michaelbprice) [reply ci/cmake-multi-platform.yml:50]: The problem is that the `g++` driver and the `gcc` driver might have different built-in include paths to search, which could cause #include directive failures if `gcc` is used when `g++` should be. I think there are also some linker implications as well (i.e., how the driver invokes `ld`).

File: ci/cmake-single-platform.yml
Comment (jgoshi) [reply ci/cmake-single-platform.yml:1]: Checked with Michael offline. We'll keep the naming convention for now. He said the good thing is that with different names, we should be able to see what mix of platforms people are targeted based on the file name, so in theory we can make a data argument to support any further change (or not).

File: ci/cmake-single-platform.yml
Comment (jgoshi) [reply ci/cmake-single-platform.yml:3]: Keeping the naming convention as is for now (see the other comment for the reasoning).

File: ci/properties/cmake-multi-platform.json
Comment (jgoshi) [reply ci/properties/cmake-multi-platform.json:2]: Will make this change and push the updates.

File: ci/properties/cmake-single-platform.json
Comment (jgoshi) [reply ci/properties/cmake-single-platform.json:2]: Same as above.

File: ci/properties/cmake-single-platform.json
Comment (jgoshi) [reply ci/properties/cmake-single-platform.json:1]: Keeping as is for now.

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:3]:
```suggestion
name: CMake on multiple platforms
```

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:16]:
```suggestion
      # Set fail-fast to false to ensure that feedback is delivered for all matrix combinations. Consider changing this to true when your workflow is stable. 
```

File: ci/cmake-multi-platform.yml
Comment (michaelbprice) [inline ci/cmake-multi-platform.yml:45]: Since you build all of the relevant matrix entries in the `include`, I think you can make this simplification as well. Obviously, you should test this suggestion.

```suggestion
      matrix:
        os: [ubuntu-latest, windows-latest]
        build_type: [Release]
        include:
          - os: windows-latest
            c_compiler: cl
            cpp_compiler: cl
          - os: ubuntu-latest
            c_compiler: gcc
            cpp_compiler: g++
          - os: ubuntu-latest
            c_compiler: clang
            cpp_compiler: clang++
```

Comment (jgoshi) [reply ci/cmake-multi-platform.yml:45]: Unfortunately, this doesn't work because include will only add a new entry if it doesn't overwrite something already added by another include. So the 2 ubuntu ones only one will apply (because only "ubuntu-latest" comes from the matrix outside the include statement)

Comment (michaelbprice) [reply ci/cmake-multi-platform.yml:45]: Yeah, you're right. I think there could be a way to make it work, but it would be really weird and probably not worth the trouble. In any manner, having an `include` and `exclude` block isn't the end of the world as it serves as a good example of how to construct that sparse matrix.

**Review by jtamsut — COMMENTED** [review]
@jgoshi We have a [PR template](https://github.com/actions/starter-workflows/blob/main/.github/pull_request_template.md) with a checklist for all new contributions to this repo. Can you go through and make sure all (relevant) items of the checklist are completed? Can you also use the PR template in your PR description? Thanks

File: ci/properties/cmake-multi-platform.json
Comment (jtamsut) [inline ci/properties/cmake-multi-platform.json:6]: nit: Add newline to end of file

File: ci/properties/cmake-single-platform.json
Comment (jtamsut) [inline ci/properties/cmake-single-platform.json:6]: same nit as above: add a newline

File: ci/properties/cmake-multi-platform.json
Comment (jtamsut) [reply ci/properties/cmake-multi-platform.json:4]: From the [GitHub docs](https://docs.github.com/en/actions/using-workflows/creating-starter-workflows-for-your-organization#creating-a-starter-workflow):

> iconName - Optional. Specifies an icon for the workflow that is displayed in the list of workflows. iconName can one of the following types:
>
>    - An SVG file that is stored in the workflow-templates directory. To reference a file, the value must be the file name without the file extension. For example, an SVG file named example-icon.svg is referenced as example-icon.
> 
>    - An icon from GitHub's set of [Octicons](https://primer.style/octicons/). To reference an octicon, the value must be octicon <icon name>. For example, octicon smiley.

File: ci/cmake-multi-platform.yml
Comment (jtamsut) [reply ci/cmake-multi-platform.yml:55]: Looking through the other workflows the convention that seems to be applied here is to add spaces before the opening and closing curly braces:

```
${{ stuff }}
```

File: ci/properties/cmake.properties.json
Comment (jtamsut) [inline ci/properties/cmake.properties.json:1]: Why did you delete this file? According to the [docs](https://docs.github.com/en/actions/using-workflows/creating-starter-workflows-for-your-organization#creating-a-starter-workflow) you need this. Am I missing something?

File: ci/cmake-multi-platform.yml
Comment (jgoshi) [reply ci/cmake-multi-platform.yml]: Made this change to add spaces.

File: ci/properties/cmake-multi-platform.json
Comment (jgoshi) [reply ci/properties/cmake-multi-platform.json]: Fixed it

File: ci/properties/cmake-single-platform.json
Comment (jgoshi) [reply ci/properties/cmake-single-platform.json]: Fixed it.

**Comment by jgoshi** [comment]
@jtamsut I went through the PR template checklist and used it in the description. Is this what you meant? Also regarding joining the GitHub Technology Partner Program I didn't do that part but work for Microsoft who I assume is a partner. Or do I need to join individually?

File: ci/properties/cmake-multi-platform.json
Comment (michaelbprice) [reply ci/properties/cmake-multi-platform.json:4]: Thanks @jtamsut. I went looking to see if there were any public domain icons that could help distinguish these workflows, but I couldn't really find any. I did find that the current CMake icon we are using has been updated, but [the version available on Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Cmake.svg) is only 200 x 200. I think we are fine skipping this for now.

**Review by michaelbprice — APPROVED** [review]
I'm approving. Pending a validation that it works on a fairly simple, but larger project, I don't think I have any further input.

I'll be away next week, so as soon as we get the required approvals from GitHub folks, I have no objections to a merge.

File: ci/cmake-multi-platform.yml
Comment (LukaszMendakiewicz) [inline ci/cmake-multi-platform.yml:73]: Should `-C` be `--build-config` in this comment?

Comment (LukaszMendakiewicz) [reply ci/cmake-multi-platform.yml:73]: And in single-platform, should the `-C` parameter be changed as well for consistency?

**Review by LukaszMendakiewicz — APPROVED** [review]

File: ci/cmake-multi-platform.yml
Comment (jgoshi) [reply ci/cmake-multi-platform.yml:73]: Thanks. I'll fix the comment. Also for the single-platform one, leaving it for now. This was what Michael had said (in another comment on this PR):
"We could probably do similar things to the contents of the single platform workflow, but let's avoid that for now so that we can avoid having multiple variables while we try to understand if users are finding the multi-platform workflow a better experience."

**Comment by jgoshi** [comment]
For validation I run both the single-platform and multi-platform workflows on a simple CMake project with a single test. I also ran the multi-platform workflow on a fork of the https://github.com/llvm/llvm-project repo (to build Clang). The Windows one passed, but both Ubuntu ones failed exceeding the max timeout of 360 minutes (both were still building). But things looked like they were working.

Per Michael's earlier comment "I'll be away next week, so as soon as we get the required approvals from GitHub folks, I have no objections to a merge". What are the next steps here in order to merge this? Thank you.

File: ci/cmake-multi-platform.yml
Comment (andrewakim) [reply ci/cmake-multi-platform.yml]: Hmmm. I don't think so and I'm not sure what we enforced in the past.

**Review by andrewakim — APPROVED** [review]

**Comment by jgoshi** [comment]
@andrewakim Do you know what the next steps are? It says "At least 1 approving review is required by reviewers with write access".

**Comment by jtamsut** [comment]
> @jtamsut I went through the PR template checklist and used it in the description. Is this what you meant? Also regarding joining the GitHub Technology Partner Program I didn't do that part but work for Microsoft who I assume is a partner. Or do I need to join individually?

Hey @jgoshi. Yeah since you are from Microsoft you don't need to become a Technology Partner. You can skip that prereq.

**Comment by jtamsut** [comment]
@jgoshi I see you listed this step as NA:

[NA] This workflow must only use actions that are produced by the language or ecosystem that the workflow supports. These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions). We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag. Additionally, workflows must include the following comment at the top of the workflow file:

```
# This workflow uses actions that are not certified by GitHub.
# They are provided by a third-party and are governed by
# separate terms of service, privacy policy, and support
# documentation.
```

Why? Can you add the blurb above to the top of the workflow file?

**Comment by jgoshi** [comment]
Sorry I put N/A by mistake. It only uses the checkout action so I'll change the checklist.

**Review by jtamsut — APPROVED** [review]

**Comment by 965022-oss** [comment]
I was newly starter workflows for Code Scanning follow up existing update.

File: ci/cmake-multi-platform.yml
Comment (swapnilrajegithubcom) [reply ci/cmake-multi-platform.yml]: You mean the should new project creation

Comment (965022-oss) [reply ci/cmake-multi-platform.yml]: It's a new action/workflow cmake-multi-platform.yml, created build type configuration using CMAKE_BUILD_TYPE now I just have to configured variables able cmake configuration connected on data application.

File: ci/cmake-multi-platform.yml
Comment (965022-oss) [reply ci/cmake-multi-platform.yml]: Workflow run with updated 
strings

**Pull request merged**
